src/pybind/mgr/insights/module.py

```diff
@@ -200,6 +200,24 @@ def _crash_history(self, hours):
 
         return result, health_check_details
 
+    def _apply_osd_stats(self, osd_map):
+        # map from osd id to its index in the map structure
+        osd_id_to_idx = {}
+        for idx in range(len(osd_map["osds"])):
+            osd_id_to_idx[osd_map["osds"][idx]["osd"]] = idx
+
+        # include stats, including space utilization performance counters.
+        # adapted from dashboard api controller
+        for s in self.get('osd_stats')['osd_stats']:
+            idx = osd_id_to_idx[s["osd"]]
```

Comment:
Is there a way this can fail? E.g. OSD was added just recently?

Reply:
@sebastian-philipp yeh it looks like that is possible in principle. i recall from talking with john spray that there isn't much (if anything) that the manager does to keep data sources in consistent states.

this code with lifted from the dashboard, so it would experience the same issue.

this pattern of grabbing some state from one API and combining it with state from another API seems pretty common. I'm wondering if it worth the effort to always handle these cases, or if it better to just use a bounded retry strategy from the caller.

Reply:
Yep, this can also fail with a `KeyError`:
```python
# in dashboard.controllers.osd
class Osd(RESTController):
    def list(self):
        osds = self.get_osd_map()
        for s in mgr.get('osd_stats')['osd_stats']:
            osds[str(s['osd'])].update({'osd_stats': s})
```

What about just catching the `IndexError` and just log the exception?

Reply:
@sebastian-philipp ok, pushed a fix that catches the key error